clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py

```diff
@@ -85,103 +119,133 @@ def __init__(self,
-                        body=body, namespace='default')
+                query_deployment_data = self._generate_config(
+                    CONFIG_FILES['query']['deployment'],
+                    image=query_image,
+                    exporter_image=CLIPPER_FRONTEND_EXPORTER_IMAGE,
+                    redis_service_host=self.redis_ip,
+                    redis_service_port=self.redis_port,
+                    cache_size=cache_size,
+                    name='query-frontend-{}'.format(query_frontend_id),
+                    id_label=str(query_frontend_id),
+                    cluster_name=self.cluster_name)
+                self._k8s_beta.create_namespaced_deployment(
+                    body=query_deployment_data, namespace=self.namespace)
 
             with _pass_conflicts():
-                body = yaml.load(
-                    open(
-                        os.path.join(cur_dir, '{}-service.yaml'.format(name))))
+                query_rpc_service_data = self._generate_config(
+                    CONFIG_FILES['query']['service']['rpc'],
+                    name='query-frontend-{}'.format(query_frontend_id),
+                    id_label=str(query_frontend_id),
+                    cluster_name=self.cluster_name)
                 self._k8s_v1.create_namespaced_service(
-                    body=body, namespace='default')
+                    body=query_rpc_service_data, namespace=self.namespace)
 
-        start_prometheus(self._k8s_v1, self._k8s_beta)
+        with _pass_conflicts():
+            query_frontend_service_data = self._generate_config(
+                CONFIG_FILES['query']['service']['query'],
+                cluster_name=self.cluster_name)
+            self._k8s_v1.create_namespaced_service(
+                body=query_frontend_service_data, namespace=self.namespace)
 
-        self.connect()
+    def _start_prometheus(self):
+        with _pass_conflicts():
+            configmap_data = self._generate_config(
+                CONFIG_FILES['metric']['config'],
+                cluster_name=self.cluster_name)
+            self._k8s_v1.create_namespaced_config_map(
+                body=configmap_data, namespace=self.namespace)
+
+        with _pass_conflicts():
+            deployment_data = self._generate_config(
+                CONFIG_FILES['metric']['deployment'],
+                version=PROM_VERSION,
+                cluster_name=self.cluster_name,
+            )
+            self._k8s_beta.create_namespaced_deployment(
+                body=deployment_data, namespace=self.namespace)
+
+        with _pass_conflicts():
+            service_data = self._generate_config(
+                CONFIG_FILES['metric']['service'],
+                cluster_name=self.cluster_name,
+            )
+            self._k8s_v1.create_namespaced_service(
+                body=service_data, namespace=self.namespace)
+
+    def _generate_config(self, file_path, **kwargs):
```

Comment:
Nice idea on using a template engine. This is much cleaner.

Reply:
This is an intermediate step. In the future we should adopt helm chart.

Reply:
Helm is definitely worth looking into. Our Kubernetes cluster is a bit more dynamic than most (as we dynamically create and delete deployments and services). The right way to do it is probably to use a Helm chart to start the initial cluster, then not use Helm for creating model deployments. This is probably worth looking into when we move to having the active API server managing containers for us.

The other thing to keep in mind is that we want to be compatible with different container orchestration systems, so we never want to only support Kubernetes.